analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/RuleChecks/ObjectsShouldNotBeDisposedMoreThanOnceBase.cs

```diff
@@ -0,0 +1,49 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2023 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using SonarAnalyzer.SymbolicExecution.Constraints;
+
+namespace SonarAnalyzer.SymbolicExecution.Roslyn.RuleChecks;
+
+public abstract class ObjectsShouldNotBeDisposedMoreThanOnceBase : SymbolicRuleCheck
+{
+    protected const string DiagnosticId = "S3966";
+    protected const string MessageFormat = "Refactor this code to make sure '{0}' is disposed only once.";
+
+    private static readonly string[] DisposeMethods = { "Dispose", "DisposeAsync", "Close" };
+
+    protected override ProgramState PreProcessSimple(SymbolicContext context)
+    {
+        var state = context.State;
+        if (context.Operation.Instance.AsInvocation() is { } invocation
+            && DisposeMethods.Contains(invocation.TargetMethod.Name))
+        {
+            if (state[invocation.Instance]?.HasConstraint(DisposableConstraint.Disposed) is true)
+            {
+                ReportIssue(context.Operation.Instance, invocation.Instance.Syntax.ToString());
```

Comment:
Let's assert the message on non-symbol expression. LIke `(a ??= b).Dispose()` or ternary `(condition ? a : b).Dispose()`

Reply:
I don't see noncompliant example of this.
I only see it the other way around, as a FN.
There should be
```
a.Dispose();
(a ??= b).Dispose(); // Noncompliant
```
and same for ternary

Reply:
Question: The other way around is still a FN since there's a path of double disposal.
Is this correct?

Reply:
Exactly

Reply:
We have the test case now. 
The original message was about asserting the message there.

Reply:
`Resource 'a ??= b' has already been disposed explicitly or implicitly through a using statement. Please remove the redundant disposal.`

Needs improvement.

Reply:
While there's a room for improvement, it's good enough given how rare the case should be